### PR TITLE
Added shadowmask and shadowmask distance support

### PR DESCRIPTION
**Pull request by lukaschod**
### Checklist for PR maker
- [ ] Have you added a backport label (if needed)? For example, the `need-backport-*` label. After you backport the PR, the label changes to `backported-*`.
- [x] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [x]  Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR. If you do add documentation, make sure to add the relevant Graphics Docs team member as a reviewer of the PR. If you are not sure which person to add, see the [Docs team contacts sheet](https://docs.google.com/spreadsheets/d/1rgUWWgwLFEHIQ3Rz-LnK6PAKmbM49DZZ9al4hvnztOo/edit#gid=1058860420).
- [x] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR
- Added shadowmask and shadowmask distance support.
- Renaming _MIXED_LIGHTING_SUBTRACTIVE into LIGHTMAP_SHADOW_MIXING. This gives us few advantages: consistency with builtin/hdrp, re-using existing builtin keyword, keyword stripping is automatically handled by core, we use keyword of core functionality.
- Refactoring lightOcclusionChannel to work as lightShadowMaskSelector (Where we used mask). Sadly I can not renamed property as it would break public API.
- Refactoring unity_ProbesOcclusion to be applied in MixRealtimeAndBakedShadows instead of light.distanceAttenuation. This way keeping similar functionality to builtin. Now light probes behaves more closely what we have in builtin.
- Changing shadow fade to be calcualte inside MixRealtimeAndBakedShadows instead of in Light gathering. As this way required by realtime and baked shadow mixing.

# **For more information please check the card https://jira.unity3d.com/browse/URP-36!**

---
### Testing status

**Manual Tests**: What did you do?
- [x] Opened test project + Run graphic tests locally
- [x] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo + Prefab overrides + Alignment in Preset
- [ ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in developer mode, you have a button at end of resources that check the paths)
- Other: 

**Comparing URP and Builtin mixed lighting. Simple scene with static walls, dynamic cube and lightprobes group.**:

*Builtin Mixed Lighting Subtractive + Probes*
- Static mesh keywords LIGHTMAP_ON LIGHTMAP_SHADOW_MIXING.
- Static mesh effected by LIGHMAP + SCREENSPACE_SHADOWS.
- Dynamic mesh keywords LIGHTMAP_SHADOW_MIXING.
- Dynamic mesh effected by PROBES + SCREENSPACE_SHADOWS.
image
*URP Mixed Lighting Subtractive + Probes*
image

*Builtin Mixed Lighting ShadowMask + Probes*
- Static mesh keywords LIGHTMAP_ON LIGHTMAP_SHADOW_MIXING SHADOWS_SHADOWMASK.
- Static mesh effected by SHADOWMASK + SCREENSPACE_SHADOWS.
- Dynamic mesh keywords LIGHTMAP_SHADOW_MIXING.
- Dynamic mesh effected by PROBES + SCREENSPACE_SHADOWS.
image
*URP Mixed Lighting ShadowMask + Probes*
image

*Builtin Mixed Lighting ShadowMask + Distance + Probes*
- Static mesh keywords LIGHTMAP_ON SHADOWS_SHADOWMASK.
- Static mesh effected by SHADOWMASK + SCREENSPACE_SHADOWS.
- Dynamic mesh keywords None.
- Dynamic mesh effected by PROBES + SCREENSPACE_SHADOWS. (Probes turn only when realtime is out of distance)
image
*URP Mixed Lighting ShadowMask + Probes* (Has harder cut-off, but it is not part of shadowmask change. However it is something we should look into.)
image

**Automated Tests**:
049_Lighting_Mixed_Subtractive
image

043_Lighting_Mixed_ShadowMask
image

134_Lighting_Mixed_ShadowMask_Distance
image

**Yamato**:
https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fshadow-mask-support

Any test projects to go with this to help reviewers?

---
### For QA
- Check if mixed lighting subtractive did not regressed in any way.
- Check if mised lighting shadowmask and shadowmask distance works similar to builtin.
- Make sure keywords are stripped if mixed lighting is not used (Either disabled in pipeline asset or not enabled in lighting settings or not baked).

### CODE REVIEW

**Review by stramit — CHANGES_REQUESTED** [review]

File: com.unity.render-pipelines.universal/Shaders/Lit.shader [hunk 1]
```diff
@@ -107,8 +108,8 @@ Shader "Universal Render Pipeline/Lit"
             #pragma multi_compile _ _ADDITIONAL_LIGHTS_VERTEX _ADDITIONAL_LIGHTS
             #pragma multi_compile_fragment _ _ADDITIONAL_LIGHT_SHADOWS
             #pragma multi_compile_fragment _ _SHADOWS_SOFT
-            #pragma multi_compile _ _MIXED_LIGHTING_SUBTRACTIVE
-            #pragma multi_compile_fragment _ _SCREEN_SPACE_OCCLUSION
+			#pragma multi_compile_fragment _ _SCREEN_SPACE_OCCLUSION
```
[the remark this answers is not in the document]
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Shaders/Lit.shader]: Fixed

File: com.unity.render-pipelines.universal/ShaderLibrary/Lighting.hlsl [hunk 2]
```diff
@@ -662,11 +678,19 @@ half3 GlobalIllumination(BRDFData brdfData, half3 bakedGI, half occlusion, half3
     return GlobalIllumination(brdfData, noClearCoat, 0.0, bakedGI, occlusion, normalWS, viewDirectionWS);
 }
 
-void MixRealtimeAndBakedGI(inout Light light, half3 normalWS, inout half3 bakedGI, half4 shadowMask)
+void MixRealtimeAndBakedGI(inout Light light, float3 positionWS, half3 normalWS, inout half3 bakedGI, half4 shadowMask)
```
[the remark this answers is not in the document]
Comment (lukaschod) [reply com.unity.render-pipelines.universal/ShaderLibrary/Lighting.hlsl:680]: Fixed

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl [hunk 3]
```diff
@@ -288,13 +312,14 @@ float3 ApplyShadowBias(float3 positionWS, float3 normalWS, float3 lightDirection
     return positionWS;
 }
 
-float ApplyShadowFade(float shadowAttenuation, float3 positionWS)
```
[the remark this answers is not in the document]
Comment (lukaschod) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:291]: Fixed

File: com.unity.render-pipelines.universal/ShaderLibrary/Lighting.hlsl [hunk 4]
```diff
@@ -662,13 +649,31 @@ half3 GlobalIllumination(BRDFData brdfData, half3 bakedGI, half occlusion, half3
...
+// Backwards compatiblity
 void MixRealtimeAndBakedGI(inout Light light, half3 normalWS, inout half3 bakedGI, half4 shadowMask)
 {
-#if defined(_MIXED_LIGHTING_SUBTRACTIVE) && defined(LIGHTMAP_ON)
+#if defined(LIGHTMAP_ON) && defined(_MIXED_LIGHTING_SUBTRACTIVE)
     bakedGI = SubtractDirectMainLightFromLightmap(light, normalWS, bakedGI);
 #endif
 }
 
+void MixRealtimeAndBakedGI(inout Light light, float3 positionWS, half3 normalWS, inout half3 bakedGI, half4 shadowMask)
```
Comment (lukaschod) [inline com.unity.render-pipelines.universal/ShaderLibrary/Lighting.hlsl:671]: Maybe we need better name for this?

**Review by stramit — APPROVED** [review]

File: com.unity.render-pipelines.universal/ShaderLibrary/Lighting.hlsl (hunk 4 above)
Comment (ellioman) [reply com.unity.render-pipelines.universal/ShaderLibrary/Lighting.hlsl:671]: Now's the chance to do it if we should.

**Review by ellioman — APPROVED** [review]

**Review by Nzollner — APPROVED** [review]

**Review by Jonasmortensen — APPROVED** [review]

**Review by thefranke — APPROVED** [review]
Nice job, looks good to me!

**Review by ellioman — APPROVED** [review]

**Review by w1shmast3r — CHANGES_REQUESTED** [review]
QA pass has revealed the following issue: **No baked shadowmask is rendered after switching the shadowmask mode to Distance shadowmask, rebaking and setting shadow distance to 0.** 
For more details regarding the repro (project, steps, repro video, and images) please refer to the following [link](https://confluence.unity3d.com/display/teamgi/Shadowmask+support+for+URP#ShadowmasksupportforURP-bug3). 
This bug needs to be fixed prior to merging this PR.

**Review by thefranke — APPROVED** [review]

**Comment by lukaschod** [comment]
> QA pass has revealed the following issue: **No baked shadowmask is rendered after switching the shadowmask mode to Distance shadowmask, rebaking and setting shadow distance to 0.**
> For more details regarding the repro (project, steps, repro video, and images) please refer to the following [link](https://confluence.unity3d.com/display/teamgi/Shadowmask+support+for+URP#ShadowmasksupportforURP-bug3).
> This bug needs to be fixed prior to merging this PR.

Fixed in latest commit

**Comment by w1shmast3r** [comment]
Added @Unity-Technologies/mobile-qa for test pass on mobiles and Linux. More information regarding platform testing could be found here: https://docs.google.com/spreadsheets/d/1OAmFn25Y4OyhQqGeZF-ZSPQPzhfWK_ZXw9-6MmgwtU8/edit#gid=868713657

**Comment by LandonTownsendUnity** [comment]
Automated test 049_Lighting_Mixed_Subtractive seems to fail right out of the gate when run manually on this branch... any idea why?

**Review by ghost — APPROVED** [review]
Switched the PBR materials in the automated tests to shadergraph materials and ensured that the results were still identical visually after baking.

**Comment by phi-lira** [comment]
What's status on this PR?

**Comment by vaidasma** [comment]
finishing manual shadowmask today

**Review by w1shmast3r — COMMENTED** [review]
Still waiting for the clarification for this recently found bug: [https://confluence.unity3d.com/display/teamgi/%5BURP%5D+-+Shadowmask+Support#id-[URP]-ShadowmaskSupport-bug7](https://confluence.unity3d.com/display/teamgi/%5BURP%5D+-+Shadowmask+Support#id-[URP]-ShadowmaskSupport-bug7) other than that the Lighting QA test pass is done.

**Review by w1shmast3r — APPROVED** [review]
Lighting QA test pass has been completed on Windows and macOS (both using Editor and Standalone Player.
Link to the feature test doc: [https://confluence.unity3d.com/display/teamgi/%5BURP%5D+-+Shadowmask+Support](https://confluence.unity3d.com/display/teamgi/%5BURP%5D+-+Shadowmask+Support)

Tested with using Unity Version 2020.2.0b3 (ef376aff2859)
Branch: trunk

Tested on the following projects:
- URP Sample scene
- CornellBox URP;
- Sponza URP.

Testing included:
- Checking that the shadowmask gets generated for all aforementioned projects;
- Testing both Shadowmask and Distance Shadowmask modes and ensuring that the behaviour is consistent with the built-in RP;
- Testing some complicated scenarios with baked shadowmask (overlapping lights, toggling shadow distance, switching between lighting modes at runtime, etc)
- Integration with different lightmap backends (CPU/GPU PLM, Enlighten), different shaders and geometry (terrains).

Test outcomes:
- Several UI and functional bugs found, addressed and verified;
- **1 known UI issue:** [No Light Overlap mode scene view visualization filter in URP](https://fogbugz.unity3d.com/f/cases/1279615/). Discussed that with Lukas and decided that due to the complexity of the potential fix, we need to fix this bug separately.
- Several unrelated issues related to URP integration with certain shaders and mixed lighting modes found. All of them have been logged to FogBugz;
- No other new regressions or ship-stoppers introduced by this PR have been found.

Approving as or revision 3f7f2fc60bb4c7d6848d458b084a8141953d6b38. Great job!

**Review by vaidasma — APPROVED** [review]
Feature looks good on Android and iOS, no artifacts or unexpected behaviours detected. Testing document: https://docs.google.com/document/d/1pshQpUoKHpe7bMmiWVCYzPrUGt44o6whiFMycNDWHIY/edit?usp=sharing

Additional project were not covered in order not to bottleneck the feature time-wise. If additional coverage needed, please contact me.

**Comment by phi-lira** [comment]
Cancelled yamato. Previous commit was green and new one just fixed a conflict in editor build settings.

**Pull request merged**